File: nodejs/src/bot/keyboards.ts
import { InlineKeyboard } from "grammy";

/**
 * 建立帶有編號列表的 Inline Keyboard，用於多選操作
 * @param items - 要顯示的項目列表（每項含 id 與顯示文字）
 * @param prefix - callback data 的前綴，例如 "keydel"
 * @returns InlineKeyboard
 *
 * 排列方式：每行一個按鈕
 * 例如：
 *  1. sk-s12ryt-xxx...  →  callback: "prefix:0"
 *  2. sk-s12ryt-yyy...  →  callback: "prefix:1"
 */
export function buildNumberedListKeyboard(
  items: { id: number; label: string }[],
  prefix: string
): InlineKeyboard {
  const kb = new InlineKeyboard();

  for (let i = 0; i < items.length; i++) {
    kb.text(`${i + 1}. ${items[i].label}`, `${prefix}:${items[i].id}`);
    if (i < items.length - 1) {
      kb.row();
    }
  }

  return kb;
}

File: python/bot/keyboards.py
"""
Keyboards for the Telegram Bot.
Inline keyboards for multi-select and single-select operations.
"""
from telegram import InlineKeyboardButton, InlineKeyboardMarkup


def make_numbered_list_keyboard(
    items: list[dict],
    label_key: str = "name",
    extra_key: str | None = None,
    callback_prefix: str = "sel",
    multi_select: bool = True,
) -> InlineKeyboardMarkup:
    """Create an inline keyboard with numbered items.

    Args:
        items: List of dicts with at least an 'id' and label_key field.
        label_key: Key to use for display text.
        extra_key: Optional extra field to show in parentheses.
        callback_prefix: Prefix for callback data.
        multi_select: If True, adds confirm button.

    Returns:
        InlineKeyboardMarkup with numbered buttons.
    """
    buttons = []
    for i, item in enumerate(items, 1):
        label = item.get(label_key, str(item.get("id", "")))
        if extra_key and item.get(extra_key):
            label = f"{label} ({item[extra_key]})"
        buttons.append([
            InlineKeyboardButton(
                f"{i}. {label}",
                callback_data=f"{callback_prefix}:{item['id']}",
            )
        ])

    if multi_select:
        buttons.append([
            InlineKeyboardButton("✅ 確認", callback_data=f"{callback_prefix}:confirm")
        ])
        buttons.append([
            InlineKeyboardButton("❌ 取消", callback_data=f"{callback_prefix}:cancel")
        ])
    else:
        buttons.append([
            InlineKeyboardButton("❌ 取消", callback_data=f"{callback_prefix}:cancel")
        ])

    return InlineKeyboardMarkup(buttons)


def make_cancel_keyboard() -> InlineKeyboardMarkup:
    """Simple cancel-only keyboard."""
    return InlineKeyboardMarkup([
        [InlineKeyboardButton("❌ 取消", callback_data="cancel")]
    ])
